src/main/java/net/minecraftforge/datafix/ForgeSchema.java

```diff
@@ -0,0 +1,368 @@
+
+        //Result map.
+        final Map<String, Type<?>> types = Maps.newConcurrentMap();
+
+        //All the core templates into which recursion of data is possible.
+        final List<TypeTemplate> templates = Lists.newArrayList();
+
+        //Check all recursion types
+        for (final Map.Entry<String, Integer> entry : RECURSIVE_TYPES.entrySet())
+        {
+            //Generate a template for the type.
+            templates.add(DSL.check(entry.getKey(), entry.getValue(), getTemplate(entry.getKey())));
+        }
+
+        //Merge them all together into one root recursive type.
+        final Optional<TypeTemplate> availableTypeTemplate = templates.stream().reduce(DSL::or);
+        if (availableTypeTemplate.isEmpty())
+        {
+            //Check for its existence or nuke.
+            return Maps.newConcurrentMap();
+        }
+        //Build a type family for our recursive types.
+        final TypeTemplate choice = availableTypeTemplate.get();
+        final TypeFamily family = new RecursiveTypeFamily(name, choice);
+
+        //Now process all other known types, taking care of recursion.
+        for (final String name : TYPE_TEMPLATES.keySet())
+        {
+            final Type<?> type;
+            final int recurseId = RECURSIVE_TYPES.getOrDefault(name, -1);
+            if (recurseId != -1)
+            {
+                type = family.apply(recurseId);
+            }
+            else
+            {
+                type = getTemplate(name).apply(family).apply(-1);
+            }
+            types.put(name, type);
+        }
+        return types;
+    }
+
+    @Override
+    public Set<String> types()
+    {
+        return TYPES.keySet();
+    }
+
+    @Override
+    public Type<?> getTypeRaw(final DSL.TypeReference type)
+    {
+        final String name = type.typeName();
+        return TYPES.computeIfAbsent(name, key -> {
+            throw new IllegalArgumentException("Unknown type: " + name);
+        });
+    }
+
+    @Override
+    public Type<?> getType(final DSL.TypeReference type)
+    {
+        final String name = type.typeName();
+        final Type<?> type1 = TYPES.computeIfAbsent(name, key -> {
+            throw new IllegalArgumentException("Unknown type: " + name);
+        });
+        if (type1 instanceof RecursivePoint.RecursivePointType<?>)
+        {
+            //The -1 is a dummy value, there can never be a type with the index of -1.
+            //The recursive point type will look for a choice type with its own index.
+            return type1.findCheckedType(-1).orElseThrow(() -> new IllegalStateException("Could not find choice type in the recursive type"));
+        }
+        return type1;
+    }
+
+    @Override
+    public TypeTemplate resolveTemplate(final String name)
+    {
+        //This is potentially invoked from the super constructor, so our field init code has not run yet.
+        //Set it to a none null value, so the super type can do its initial construction.
```

Comment:
If this is actually invoked by the super constructor doesn't that mean this method will always throw an `IllegalArgumentException` then as the current logic for being invoked early via the super constructor is:
- initialize to empty map
- get from empty map if not present (which is always the case here) throw IAE

Reply:
But given it will always error anyway, wouldn't it make more sense to throw an `IllegalStateException` there rather than initialize it.